### PR TITLE
Service ctor exception crash

### PR DESCRIPTION
**Pull request by carneyweb**
If for some reason an exception is thrown in the constructor of a service implementation, a later call to ServiceObjectsBase::GetService() would throw an exception because the ServiceInterfaceMap is not populated for this service. Added a null check against the interfacemap to avoid this exception.

### CODE REVIEW

File: compendium/DeclarativeServices/test/TestComponentLifecycle.cpp [hunk 1]
```diff
@@ -73,11 +74,14 @@ TEST_F(tServiceComponent, testThrowingLifeCycleHooks) //DS_TOI_9
 {
   auto testBundle = StartTestBundle("TestBundleDSTOI9");
   auto ctxt = framework.GetBundleContext();
+  auto const refs = ctxt.GetServiceReferences<test::LifeCycleValidation>();
```
Comment (jeffdiclemente) [inline compendium/DeclarativeServices/test/TestComponentLifecycle.cpp]: is this necessary? i don't see ```refs``` being used anywhere.

File: compendium/test_bundles/TestBundleDSTOI9/src/ServiceImpl.cpp [hunk 2]
```diff
@@ -1,15 +1,26 @@
...
-  {
-    deactivated = true;
-    throw std::runtime_error("Exception thrown from user code");
-  };
-  ServiceComponent9::~ServiceComponent9() {};
+
+ServiceComponent9::ServiceComponent9()
+  : test::LifeCycleValidation()
+{
+  throw std::exception {};
```
Comment (jeffdiclemente) [inline compendium/test_bundles/TestBundleDSTOI9/src/ServiceImpl.cpp]: you should be able to test for this condition without DS by creating a ```ServiceFactory``` derived class which throws an exception when its constructed.

File: framework/src/service/ServiceObjects.cpp [hunk 3]
```diff
@@ -138,20 +138,17 @@ std::shared_ptr<void> ServiceObjectsBase::GetService() const
     return nullptr;
   }
 
-  // interfaceMap can be null under certain circumstance
-  // e.g. if a constructor is thrown in the service implementation.
-  // In that case, we bail out early.
   auto interfaceMap = d->GetServiceInterfaceMap();
+  // g2113391: interfaceMap can be null under certain circumstance e.g. if a constructor
+  // is thrown in the service implementation.  In that case, we bail out early.
```
Comment (jeffdiclemente) [inline framework/src/service/ServiceObjects.cpp]: you'll want to remove the issue #. its not relevant here.

File: framework/test/gtest/ServiceFactoryTest.cpp [hunk 4]
```diff
@@ -140,15 +159,41 @@ TEST_F(ServiceFactoryTest, TestGetServiceThrows)
...
   context.RemoveListener(std::move(lToken));
 }
 
+// Tests the return type and FrameworkEvent generated when a
+// ServiceFactory instance throws an exception in it's GetService
+// callback
+TEST_F(ServiceFactoryTest, TestGetServiceObjectThrows)
+{
+  auto context = framework.GetBundleContext();
+  auto sf = std::make_shared<MockFactory>();
```
Comment (jeffdiclemente) [inline framework/test/gtest/ServiceFactoryTest.cpp]: shouldn't this be creating an instance of ```ThrowingFactory```?

File: framework/test/gtest/ServiceFactoryTest.cpp [hunk 5]
```diff
@@ -140,15 +143,42 @@ TEST_F(ServiceFactoryTest, TestGetServiceThrows)
...
+                                                   Constants::SERVICE_SCOPE
+                                                   , Any(Constants::SCOPE_PROTOTYPE)
+                                                 }});
+
+  auto sref = context.GetServiceReference<ITestServiceA>();
+  auto serviceObjects = context.GetServiceObjects<ITestServiceA>(sref);
+
+  // Next line used to crash if a service implementation threw an exception in the
+  // constructor. This test case checks to make sure that the fix is in place.
+  (void)serviceObjects.GetService();
```
Comment (jeffdiclemente) [inline framework/test/gtest/ServiceFactoryTest.cpp]: probably want to wrap this in a ```EXPECT_NO_THROW``` so gtest can flag it as a failure.

Comment (carneyweb) [reply framework/test/gtest/ServiceFactoryTest.cpp]: Done.

File: framework/src/service/ServiceObjects.cpp [hunk 6]
```diff
@@ -138,12 +138,18 @@ std::shared_ptr<void> ServiceObjectsBase::GetService() const
...
   }
 
-  std::shared_ptr<UngetHelper> h(
-    new UngetHelper(d->GetServiceInterfaceMap(),
-                    d->m_reference,
-                    d->m_context->bundle->shared_from_this()));
-  return std::shared_ptr<void>(
-    h, (h->interfaceMap->find(d->m_reference.GetInterfaceId()))->second.get());
+  auto interfaceMap = d->GetServiceInterfaceMap();
+  // interfaceMap can be null under certain circumstance e.g. if a constructor is thrown
```
Comment (jeffdiclemente) [inline framework/src/service/ServiceObjects.cpp]: nitpick: a constructor is not thrown, it throws an exception.

**Pull request merged**